### DISCUSSION TITLE
Dynamic valueEditors based on operators for a field

### DISCUSSION THREAD

**Discussion by muddana-harini** [react-querybuilder/react-querybuilder#571] [Q&A] [answered]
Is there a way to get different value editors like dropdown or textbox dynamically based on the selected operator for a field?? If yes, how?

**Comment by jakeboone02** [comment, marked as answer]
This should get you pointed in the right direction:

https://react-querybuilder.js.org/docs/tips/custom-with-fallback

**Reply by muddana-harini** [reply]
Yes, this is based on fields right, but I want to implement it based on the operator. Does that work for the operator too?
For example, if we give gender as a field, if I select "in" operator it should show a dropdown and for the other operator, it should show a textbox.
is this possible?

**Reply by jakeboone02** [reply]
Yeah, same principle. An `operator` prop is always passed to the value editor, so instead of switching editors based on `props.fieldData.datatype` like the example, use `props.operator`.

**Reply by muddana-harini** [reply]
thank you, this helped me but it is only getting applicable for first group of rules, why?

```tsx
import React, { useState, useEffect } from 'react';
import { QueryBuilder, formatQuery} from 'react-querybuilder';

const App = () => {

  const inputTypeOptions = [
    { name: 'dropDown1', label: 'Drop Down 1' },
    { name: 'dropDown2', label: 'Drop Down 2' },
    { name: 'dropDown3', label: 'Drop Down 3' },
  ];

  const fields = [
    { name: 'firstName', label: 'First Name', datatype: 'text' },
    { name: 'lastName', label: 'Last Name', datatype: 'text' },
    { name: 'birthday', label: 'Birthday', datatype: 'date', inputType: 'date' },
    {
      name: 'inputType',
      label: 'Input Type',
    },
    // Define your fields here
  ];

  const valueSources = {
    inputType: {
      label: 'Input Type',
      type: 'select',
      values: inputTypeOptions.map(option => ({ name: option.name, label: option.label })),
    },
  };

  const getOperators = (fieldName) => {
    switch (fieldName) {
      case 'firstName':
        return [
          { name: 'beginsWith', label: 'Begins With' },
          { name: 'contains', label: 'Contains' },
          // Add other text-specific operators as needed
        ];
      case 'lastName':
        return [
          { name: 'in', label: 'In' },
          { name: 'notIn', label: 'Not In' },
          // Add other text-specific operators as needed
        ];
      case 'birthday':
        return [
          { name: '=', label: 'On' },
          { name: '!=', label: 'Not On' },
          { name: '<', label: 'Before' },
          { name: '<=', label: 'On or Before' },
          { name: '>', label: 'After' },
          { name: '>=', label: 'On or After' },
          { name: 'between', label: 'Between' },
        ];
      case 'inputType':
        return [
          { name: '=', label: 'On' },
          { name: '!=', label: 'Not On' },
          { name: '>', label: 'greaterthan' },
          { name: '<', label: 'less than' }
        ];
      default:
        return [];
    }
  };

  const combinatorOptions = [
    { name: 'and', label: 'AND' },
    { name: 'or', label: 'OR' },
    { name: 'none', label: 'NONE' },
  ];

  const [query, setQuery] = useState(initialQuery);
  const [savedQuery, setSavedQuery] = useState({});
  const [updatedFields, setUpdatedFields] = useState(fields)

  const handleQueryChange = (newQuery) => {
    if (newQuery && newQuery.rules) {
      newQuery.rules.forEach((rule) => {
        if (rule.operator) {
          console.log(`Operator selected for field ${rule.field}: ${rule.operator}`);
          if (rule.operator === '>' || rule.operator === '<') {
            const updatedFields = fields.map((field) => {
              if (field.name === rule.field) {
                return {
                  ...field,
                  valueEditorType: null,
                  value: null,
                  datatype: 'text'
                };
              }
              return field;
            });
            setUpdatedFields(updatedFields);
          }
          else if (rule.operator === '=' || rule.operator === '!=') {
            const updatedFields = fields.map((field) => {
              if (field.name === rule.field) {
                return {
                  ...field,
                  valueEditorType: 'select',
                  values: inputTypeOptions
                };
              }
              return field;
            });
            setUpdatedFields(updatedFields);
          }

        }
      });
    }
    setQuery(newQuery);
  };

  const handleSaveClick = () => {
    if (!query || !query.rules || query.rules.length === 0) {
      alert('Query is empty. Please add conditions.');
    } else if (
      fields.some(field => field.datatype === 'text' && !query.rules.some(rule => rule.field === field.name && rule.value))
    ) {
      alert('Text input fields cannot be empty.');
    } else {
      saveDataToServer(query);
    }
  };


  const saveDataToServer = (data) => {
    fetch('http://localhost:4001/save-data', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
      .then((response) => response.json())
      .then(() => {
        alert('Query saved successfully!');
      },
        setSavedQuery(data)
      )
      .catch((error) => {
        console.error('Error saving query:', error);
      });
  };

  useEffect(() => {
    // Fetch initial data from the server
    fetch('http://localhost:4001/get-data')
      .then((response) => response.json())
      .then((fetchedData) => {
        setQuery(fetchedData);
        setSavedQuery(fetchedData);
      })
      .catch((error) => {
        console.error('Error:', error);
      });
  }, []);

  return (
    <div>
      <h1>Benefit Rules</h1>
      <QueryBuilder
        fields={updatedFields}
        query={query}
        onQueryChange={handleQueryChange}
        getOperators={getOperators}
        combinators={combinatorOptions}
        controlClassnames={{ queryBuilder: 'queryBuilder-branches' }}
        valueSources={valueSources}
        validator={defaultValidator}
      />

      <button onClick={handleSaveClick}>Save</button>

      <h4>Query</h4>
      <pre>
        <code>{formatQuery(savedQuery, 'json')}</code>
      </pre>
    </div>
  );

};

export default App;
```

I added my own validations, even they are being applied only to the first group.

**Reply by jakeboone02** [reply]
You're processing the `query.rules` array as a flat list of rules instead of recursively. See an example of recursive processing [here](https://react-querybuilder.js.org/docs/tips/adding-removing-query-properties#adding-properties) (expand the "Manual recursion" section at the bottom).

I refactored your validation code to work recursively in [this sandbox](https://codesandbox.io/p/sandbox/competent-haze-rjjdhf), and the code is copied below. (It's converted to TypeScript because I needed to make sure everything was correctly shaped.)

I also did some more "code review" stuff that should help, like moving unchanging constants outside of the component function. All my changes are indicated with comments beginning with `/** JB: `.

```tsx
import { useState, useEffect } from 'react';
import { QueryBuilder, defaultValidator, formatQuery, RuleGroupType, RuleType, isRuleGroup} from 'react-querybuilder';
/** JB: ADDED NEXT LINE FOR STYLING */
import 'react-querybuilder/dist/query-builder.scss'

/** JB: MOVED OUT OF COMPONENT FUNCTION */
const inputTypeOptions = [
  { name: 'dropDown1', label: 'Drop Down 1' },
  { name: 'dropDown2', label: 'Drop Down 2' },
  { name: 'dropDown3', label: 'Drop Down 3' },
];

/** JB: MOVED OUT OF COMPONENT FUNCTION */
const fields = [
  { name: 'firstName', label: 'First Name', datatype: 'text' },
  { name: 'lastName', label: 'Last Name', datatype: 'text' },
  { name: 'birthday', label: 'Birthday', datatype: 'date', inputType: 'date' },
  { name: 'inputType', label: 'Input Type' },
  // Define your fields here
];

/**
 * JB: COMMENTED THIS
 * I'm not sure what you were going for here.
 * Value source docs: https://react-querybuilder.js.org/docs/components/querybuilder#getvaluesources
 */
// const valueSources = {
//   inputType: {
//     label: 'Input Type',
//     type: 'select',
//     values: inputTypeOptions.map(option => ({ name: option.name, label: option.label })),
//   },
// };

/** JB: MOVED OUT OF COMPONENT FUNCTION */
const getOperators = (fieldName: string) => {
  switch (fieldName) {
    case 'firstName':
      return [
        { name: 'beginsWith', label: 'Begins With' },
        { name: 'contains', label: 'Contains' },
        // Add other text-specific operators as needed
      ];
    case 'lastName':
      return [
        { name: 'in', label: 'In' },
        { name: 'notIn', label: 'Not In' },
        // Add other text-specific operators as needed
      ];
    case 'birthday':
      return [
        { name: '=', label: 'On' },
        { name: '!=', label: 'Not On' },
        { name: '<', label: 'Before' },
        { name: '<=', label: 'On or Before' },
        { name: '>', label: 'After' },
        { name: '>=', label: 'On or After' },
        { name: 'between', label: 'Between' },
      ];
    case 'inputType':
      return [
        { name: '=', label: 'On' },
        { name: '!=', label: 'Not On' },
        { name: '>', label: 'greaterthan' },
        { name: '<', label: 'less than' },
      ];
      default:
        return [
          /**
           * JB: ADDED NEXT LINE
           * Operator list shouldn't be empty.
           */
          { name: '=', label: 'Is' }
        ];
  }
};

/** JB: MOVED OUT OF COMPONENT FUNCTION */
const combinatorOptions = [
  { name: 'and', label: 'AND' },
  { name: 'or', label: 'OR' },
  { name: 'none', label: 'NONE' },
];

/**
 * JB: ADDED NEXT LINE
 * This variable wasn't defined in the code you provided
 */
const initialQuery: RuleGroupType = { combinator: 'and', rules: []};

export const App = () => {
  const [query, setQuery] = useState(initialQuery);
  /** JB: CHANGED FROM `{}` TO A PROPER QUERY OBJECT */
  const [savedQuery, setSavedQuery] = useState(initialQuery);
  const [updatedFields, setUpdatedFields] = useState(fields)
  
  /** JB: REFACTORED TO PROCESS QUERY RECURSIVELY */
  const handleQueryChange = (newQuery: RuleGroupType) => {
    const processRule = (rule: RuleType) => {
      if (rule.operator) {
        console.log(`Operator selected for field ${rule.field}: ${rule.operator}`);
        if (rule.operator === '>' || rule.operator === '<') {
          const updatedFields = fields.map((field) => {
            if (field.name === rule.field) {
              return {
                ...field,
                valueEditorType: null,
                value: null,
                datatype: 'text'
              };
            }
            return field;
          });
          setUpdatedFields(updatedFields);
        }
        else if (rule.operator === '=' || rule.operator === '!=') {
          const updatedFields = fields.map((field) => {
            if (field.name === rule.field) {
              return {
                ...field,
                valueEditorType: 'select',
                values: inputTypeOptions
              };
            }
            return field;
          });
          setUpdatedFields(updatedFields);
        }
      }
    }

    const processGroup = (rg: RuleGroupType) => {
      rg.rules.forEach(r => {
        if (isRuleGroup(r)) {
          processGroup(r);
        } else {
          processRule(r);
        }
      })
    };

    processGroup(newQuery);

    setQuery(newQuery);
  };

  const handleSaveClick = () => {
    if (!query || !query.rules || query.rules.length === 0) {
      alert('Query is empty. Please add conditions.');
    } else if (
      fields.some(field => field.datatype === 'text' && !(query.rules as RuleType[]).some(rule => rule.field === field.name && rule.value))
    ) {
      alert('Text input fields cannot be empty.');
    } else {
      saveDataToServer(query);
    }
  };

  const saveDataToServer = (data: RuleGroupType) => {
    fetch('http://localhost:4001/save-data', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
      .then((response) => response.json())
      .then(() => {
        /** JB: MOVED THIS LINE HERE FROM AFTER THE CLOSING BRACE */
        setSavedQuery(data)
        alert('Query saved successfully!');
      },
      /**
       * JB: THIS IS WHERE THE LINE ABOVE WAS
       * This was invalid Promise.then syntax,
       * so I assumed you wanted it inside the
       * onresolve handler.
       */
      )
      .catch((error) => {
        console.error('Error saving query:', error);
      });
  };

  useEffect(() => {
    // Fetch initial data from the server
    fetch('http://localhost:4001/get-data')
      .then((response) => response.json())
      .then((fetchedData) => {
        setQuery(fetchedData);
        setSavedQuery(fetchedData);
      })
      .catch((error) => {
        console.error('Error:', error);
      });
  }, []);

  return (
    <div>
      <h1>Benefit Rules</h1>
      <QueryBuilder
        fields={updatedFields}
        query={query}
        onQueryChange={handleQueryChange}
        getOperators={getOperators}
        combinators={combinatorOptions}
        controlClassnames={{ queryBuilder: 'queryBuilder-branches' }}
        /**
         * JB: COMMENTED THIS
         * See `valueSources` note above
         */
        // valueSources={valueSources}
        validator={defaultValidator}
      />

      <button onClick={handleSaveClick}>Save</button>

      <h4>Query</h4>
      <pre>
        {/** JB: CHANGED `savedQuery` TO `query` */}
        <code>{formatQuery(query, 'json')}</code>
      </pre>
    </div>
  );
};
```

**Reply by dheerajkumar10** [reply]
Hi @jakeboone02 , I have been following this chain for the last week, I tried flattening and changing the arrays, the rule group will be extracted and operator logic will be applied to the inputtypes but if we have 2  inputtypes in 2 different groups always last input type operator will change other types like if we have ">" in the first inputtype and "=" in the second operator  "=" will change the datatype of first inputytpe also because here we are changing the fields directly is there any way where we can just change the type of that particular element so that it will not change the fields if we update fields it changes for all the elements with same type.

cc: @muddana-harini

**Reply by jakeboone02** [reply]
I'm sorry @dheerajkumar10, I can't really grok what you're asking. Can you provide some example code or better yet a working example that demonstrates your issue?

**Reply by dheerajkumar10** [reply]
Thanks @jakeboone02 for the fast reply if you check the code sample you shared on sandbox try to add a rule group with inputtype and add one more group and add one more input type rule and change the operator of the second input type to > or < it will change the first inputtype dropdown to the text box. this is because we are updating the fields for the whole inputtype field.

https://codesandbox.io/p/sandbox/competent-haze-rjjdhf?file=%2Fsrc%2FApp.tsx%3A6%2C43

**Reply by jakeboone02** [reply]
That's almost certainly an issue with the custom `handleQueryChange` function.

**Reply by dheerajkumar10** [reply]
Yes, is there any way we can change the data type for that particular element instead of changing the data type of particular field so that we can avoid change for other elements of similar type

**Reply by muddana-harini** [reply]
Hi @jakeboone02 , Is there anyway you could help us solve the issue?

**Reply by jakeboone02** [reply]
I had a closer look at your code and I think I can see what you're trying to do. Instead of updating the `fields` array itself based on the `operator` selected, you need to use the `getValues` and `getValueEditorType` callbacks to dynamically choose the properties based on the `operator`.

I've updated the sandbox with these recommended changes. Pay close attention to lines 180-195.

https://codesandbox.io/p/sandbox/competent-haze-rjjdhf?file=%2Fsrc%2FApp.tsx%3A180%2C1-196%2C1

Relevant documentation:
- [`getValues`](https://react-querybuilder.js.org/docs/components/querybuilder#getvalues)
- [`getValueEditorType`](https://react-querybuilder.js.org/docs/components/querybuilder#getvalueeditortype)

**Reply by dheerajkumar10** [reply]
I @jakeboone02, Thanks for the reply. This is working for me. Nice to connect with you and solve the problem, I didn't get the idea using [getValueEditorType.
Thanks again.

**Reply by fxylee** [reply]
A few years ago we solve it by a component like `ValueEditorDispatcher`.

And now, we have a `ValueEditor` to build json value

**Reply by koseenes** [reply]
I need to use dynamically change combobox select type, when i choose "=" or "!=" operator select control have to be single select, when i choose "in" or "not in" select control have to multiselect. 

This is sample code: https://codesandbox.io/p/sandbox/cocky-goldberg-nqrshg?file=%2Fsrc%2FApp.tsx%3A67%2C2-77%2C5

I expect below getValueEditorType function change ValueEditorType dynamically, but it didnt change. 

 ```
const getValueEditorType = (fieldName: string, operator: string) => {
    if (fields.find(f => f.name === fieldName)!.valueEditorType == 'select') {
      if (operator === 'in' || operator === 'notIn') return 'multiselect';
      else return 'select';
    }
    return (
      (fields.find(f => f.name === fieldName)!
        .valueEditorType as ValueEditorType) ?? 'text'
    );
  };
```

Do you have any recommendation?

**Reply by jakeboone02** [reply]
@koseenes The `valueEditorType` property of the field will supersede the return value of `getValueEditorType`. Fortunately `valueEditorType` can be a function that takes an operator, so you can apply the same logic to the field itself like this:

```ts
const getValueEditorType = (operator: string) =>
  operator === 'in' || operator === 'notIn' ? 'multiselect' : 'select';

const fields: Field[] = [
  {
    name: 'inputType',
    label: 'Input Type',
    valueEditorType: getValueEditorType,
    values: inputTypeOptions,
  },
];
```

I forked and corrected your sandbox here: https://codesandbox.io/p/sandbox/clever-dew-36c8lf?file=%2Fsrc%2FApp.tsx%3A17%2C1-28%2C1

**Reply by koseenes** [reply]
Thank you @jakeboone02, resolved.

**Comment by muddana-harini** [comment]
Hey Jake,
                I am trying to get the data from api. it is consoling the value  but returning the data in dropdown.
                instead if i directly send the data hardcoded without calling api it is showing the dropdown.
                
                getValues={async (field, op) => {
                    try {
                    if (op === 'isInList') {
              
                        console.log("Field:",field)
                        const data = await fetchData(field);
                        console.log("DATA:", data);
                        return data;
                      } 
                      return [];
                    }catch (error) {
                        console.error("Error fetching data:", error);
                        return [];
                      }
                  }}
                  
                  can you help me with this.?

**Reply by jakeboone02** [reply]
RQB doesn't support `async` functions in any props. I would recommend a library like [`react-select`](https://react-select.com/) if you want to fetch values from an API.

**Reply by muddana-harini** [reply]
Thank you for the reply, I managed to get the dropdown but I found out other issue. 
if my filed has operator list is having dropdown operator in the first option
      "drgCode": [
        { "name": "isInList", "label": "Is in List", "valueEditorType": "select" },
        { "name": "eq", "label": "Equal to", "inputType": "text" },
        { "name": "isInSet", "label": "Is In Set of", "inputType": "text" }
      ]
then it is taking the first value automatically in the query or taking the selected value. but if the list is like
      "drgCode": [
        { "name": "eq", "label": "Equal to", "inputType": "text" },
        { "name": "isInSet", "label": "Is In Set of", "inputType": "text" }
        { "name": "isInList", "label": "Is in List", "valueEditorType": "select" },
      ]
then it is not taking the first value if selected....to select the first value i had to select any other option in the dropdown and then again select first value to get that in the query.
Any idea why this is happening..or any clue to solve this issue.

**Reply by jakeboone02** [reply]
Please replicate the issue in CodeSandbox or similar live example. You can start [one of the example templates](https://github.com/react-querybuilder/react-querybuilder/tree/main/examples/basic-ts).

**Comment by dheerajkumar10** [comment]
Hey @jakeboone02, thank you for the reply, I think what @muddana-harini  is trying to tell is when we change the field type based on operator, like we change from textbox to select box the value "dropdown" is not automatically applied in value field. I changed the code and added it here
added console logs for reference.

If we have select option as default one then its picking the value

try changing 

```
   case 'inputType':
      return [
        { name: '>', label: 'greaterthan' },
        { name: '=', label: 'On' },
        { name: '!=', label: 'Not On' },
        { name: '<', label: 'less than' },
      ];
```
to
```
      case 'inputType':
      return [
        { name: '=', label: 'On' },
        { name: '>', label: 'greaterthan' },
        { name: '!=', label: 'Not On' },
        { name: '<', label: 'less than' },
      ];
```

you can see when its default select for inputtype it picks first option as value but when we change it default to text and then change the operator to pick select it doesn't pick value even when you click in that first option (you have to click other option to make change detection work) just check the query value  
      
https://codesandbox.io/p/devbox/relaxed-shape-yhftfh?file=%2Fsrc%2FApp.tsx%3A68%2C47

**Reply by jakeboone02** [reply]
Try adding the [`resetOnOperatorChange` prop](https://react-querybuilder.js.org/docs/components/querybuilder#resetonoperatorchange).

**Reply by dheerajkumar10** [reply]
Thanks @jakeboone02.

**Comment by muddana-harini** [comment]
Hey Jake, 
When I tried giving the "between" operator for the "date" datatype it is not showing two input boxes. Whereas, for "text" and "number" datatype it is showing two input boxes....only for date datatype it is showing only one box I am calling operators from separate json file, is that the reason?

const getOperators = (fieldName) => {
  if (fieldName in operators) {
    return operators[fieldName];
  }
  return [];
};
  const handleQueryChange = (newQuery) => {
    processGroup(newQuery);
    setQuery(newQuery);
  };

json file--->
{
    "fields": [
      { "name": "birthDate", "label": "Birth Date", "datatype": "date", "inputType": "date" },
    ],
    "operators": {
      "birthDate": [
        { "name": "between", "label": "Between", "inputType": "date" }
      ]
    }
  }

**Reply by jakeboone02** [reply]
Can you put together a sandbox or minimal repo?

**Reply by muddana-harini** [reply]
https://codesandbox.io/p/sandbox/github/muddana-harini/querybuilder/tree/main/

please check accident date and service date with between operator...it will be showing only one box fir input

**Reply by jakeboone02** [reply]
@muddana-harini It looks like you've mixed up `valueEditorType` with `inputType`.

This sandbox fixes the issue: https://codesandbox.io/p/devbox/4s7kfg?file=%2Fsrc%2FApp.jsx%3A183%2C1-184%2C1

The change was this, in the `getValueEditorType` prop:

```diff
-    return selectedField.inputType || "text";
+    return selectedField.valueEditorType || "text";
```

**Reply by muddana-harini** [reply]
Thank you...it worked

**Comment by muddana-harini** [comment]
Hey Jake, 
Hope you are doing good. I have 2 queries.
1) I am looking at your multi-select component but I have a big list of values to display so preferring to use multi-select from dropdown do you have any example implementations. 
2) And I want to display two input values for paritulcar fields...that is when I select isinlist for particular fields I may want to show both dropdown select box and a textbox to write some other value...I have gone through some discussion but wanted to check if you can guide me through any simple solution or these. 

Hoping to hear from you soon. 
Thank you.

https://codesandbox.io/p/sandbox/adoring-lovelace-tsxfks?file=%2Fsrc%2FApp.js

**Reply by jakeboone02** [reply]
1. As discussed on Discord:

> The website landing page has a demo of `react-select`, which does a nice drop-down multiselect. The code for the website demo is [here](https://github.com/react-querybuilder/react-querybuilder/blob/708896cfd10c33285ae7ab72d6343a9bd52dcb00/website/src/pages/index.tsx#L73-L130), although it's a little over-complicated by itself since it's meant to integrate with the docusaurus theme and some other stuff.

2. Generally I recommend storing both values in the `value` property, either as an array or delimited string, and rendering two controls with a custom value editor. When one control updates, you need to call `props.handleOnChange` with the updated value _plus_ the value that was not updated, not just the updated value because you will lose the value of the other control.

**Comment by muddana-harini** [comment]
> 1. As discussed on Discord:
> 
> > The website landing page has a demo of `react-select`, which does a nice drop-down multiselect. The code for the website demo is [here](https://github.com/react-querybuilder/react-querybuilder/blob/708896cfd10c33285ae7ab72d6343a9bd52dcb00/website/src/pages/index.tsx#L73-L130), although it's a little over-complicated by itself since it's meant to integrate with the docusaurus theme and some other stuff.
> 
> 2. Generally I recommend storing both values in the `value` property, either as an array or delimited string, and rendering two controls with a custom value editor. When one control updates, you need to call `props.handleOnChange` with the updated value _plus_ the value that was not updated, not just the updated value because you will lose the value of the other control.

for 2nd solution -->
its not about binding values to output. I am not able to customize value editor to have two input fields one selectbox and one text box for selected field and operator. That is for example ...if I select a field and operator which shows a list of dropdown I also need a text box to enter a integer value in the same row.

**Reply by jakeboone02** [reply]
A custom value editor can render anything you want. Here's an example that renders a `<select>` and an `<input>` for a particular `field`, with a fallback to the default value editor:

[Working sandbox](https://codesandbox.io/p/devbox/adoring-khayyam-3gdrcm?file=%2Fsrc%2FApp.tsx%3A23%2C1-46%2C1)

```tsx
const MyValueEditor = (props: ValueEditorProps) => {
  if (props.field === 'firstName') {
    const [title = 'mr', name = ''] = props.value?.split('|') ?? ['', ''];
    return (
      <div className={props.className}>
        <select
          value={title}
          onChange={e => props.handleOnChange(`${e.target.value}|${name}`)}>
          <option value="mr">Mr.</option>
          <option value="ms">Ms.</option>
          <option value="mrs">Mrs.</option>
          <option value="mx">Mx.</option>
        </select>
        <input
          type="text"
          value={name}
          onChange={e => props.handleOnChange(`${title}|${e.target.value}`)}
        />
      </div>
    );
  }
  return <ValueEditor {...props} />;
};
```
